include/MatrixClient.h

```diff
@@ -94,9 +94,9 @@ public slots:
                              const QString &homeserver,
                              const QString &token);
         void versionSuccess();
-        void imageUploaded(const QString &roomid, const QString &filename, const QString &url);
-        void fileUploaded(const QString &roomid, const QString &filename, const QString &url);
-        void audioUploaded(const QString &roomid, const QString &filename, const QString &url);
+        void imageUploaded(const QString &roomid, QSharedPointer<QFile> file, const QString &url);
+        void fileUploaded(const QString &roomid, QSharedPointer<QFile> file, const QString &url);
+        void audioUploaded(const QString &roomid, QSharedPointer<QFile> file, const QString &url);
```

Comment:
Why are these `QFile`? The whole point of the change to use `QIODevice` is that the uploaded data is not necessarily a file. As the filename (and only the filename) is needed for UI purposes, this should remain a `QString`.

Reply:
Ah this is an oversight on my end. You are correct, they do not need to take a `QFile`. However, it is not clear to me what the `QString` should be, if the data does not come from a file (where the string would obviously be the file name).

Reply:
I'd just invent something descriptive, depending on the actual origin of the data. For example, this PR might use the name "clipboard.png" for UI and upload filename purposes.

Reply:
Just to clarify, how would I know if I received a file or a buffer containing data? Would the best method be to try to cast to a `QFile` and use the file name if it exists, and if the cast fails, then use the "clipboard.png" mentioned?

Reply:
You should not need to know what specific `QIODevice` generic code receives; doing so defeats the purpose of having generic code. For example, the approach you outline will break down once there are non-clipboard sources of in-memory data (e.g. imagery captured directly from a camera). Justifiable uses of dynamic casting in C++ are extremely rare.

Instead, whenever a filename will eventually be needed alongside a `QIODevice`, you should pass a filename along from the very beginning. For example, the signature of `FilteredTextEdit::image` could be changed to `void FilteredTextEdit::image(QSharedPointer<QIODevice> data, const QString &filename)`